app/assets/javascripts/discourse/app/static/prosemirror/core/serializer.js

```diff
@@ -0,0 +1,38 @@
+import {
+  defaultMarkdownSerializer,
+  MarkdownSerializer,
+} from "prosemirror-markdown";
+
+export default class Serializer {
+  #pmSerializer;
+
+  constructor(extensions, includeDefault = true) {
+    this.nodes = includeDefault ? { ...defaultMarkdownSerializer.nodes } : {};
+    this.nodes.hard_break = (state) => state.write("\n");
+
+    this.marks = includeDefault ? { ...defaultMarkdownSerializer.marks } : {};
+
+    this.#extractNodeSerializers(extensions);
+    this.#extractMarkSerializers(extensions);
+
+    this.#pmSerializer = new MarkdownSerializer(this.nodes, this.marks);
+  }
+
+  convert(doc) {
+    return this.#pmSerializer.serialize(doc);
+  }
+
+  #extractNodeSerializers(extensions) {
+    return extensions.reduce((acc, { serializeNode }) => {
+      Object.assign(acc, serializeNode);
+      return acc;
+    }, this.nodes);
```

Comment:
(non-blocking personal opinion)

I am not keen on how frequently `reduce()` is being used in this PR. e.g. for this case, I think this is 100x more readable, and achieves the same result:

```suggestion
    for (const { serializeNode } of extensions){
      Object.assign(this.nodes, serializeNode);
    }
```